File: ios/Godstone/Sources/LabMesh/LabMeshApp.swift
import Foundation
import SwiftUI
import GodstoneMesh
import GodstoneCore

// ---------------------------------------------------------------------------
// T54 -- the LabMesh iOS target's application seam.
//
// This target carrieth the lab's OWN bundle identity (io.godstone.labmesh,
// declared in ios/project.yml) and nothing else: the runtime it driveth is the
// canonical one, reached through `LabRuntime`, which liveth in the nonshipping
// GodstoneMesh module. There is no lab-only copy of any authority, and no way to
// manufacture readiness from here.
//
// It is VISIBLY EXPERIMENTAL: the entry point sayeth so, the bundle id sayeth so,
// and the profile gate asserteth that this target and its sources can never reach
// the LIGHT Archive-only application.
// ---------------------------------------------------------------------------

/// The lab application's own marker: identity and profile a reader -- and the
/// profile gate -- can inspect without starting a runtime.
public enum LabMeshApp {
    public static let applicationId: String = LabProfile.labBundleId
    public static let shippingApplicationId: String = LabProfile.shippingBundleId
    public static let profile: String = LabProfile.name
    public static let experimental: Bool = LabProfile.experimental
    public static let manufacturesReadiness: Bool = LabProfile.manufacturesReadiness

    /// Compose the real runtime the lab driveth.
    public static func compose() throws -> LabRuntime { try LabRuntime.compose() }

    /// The honest readiness statement (all platform fields false).
    public static func readiness() -> LabReadiness { LabRuntime.readinessStatement() }
}

/// The lab's own view. It existeth so the target buildeth as an application; the
/// interesting surface is `LabMeshApp`, which carrieth no readiness setter.
public struct LabMeshRootView: View {
    public init() {}

    public var body: some View {
        VStack(alignment: .leading, spacing: 8) {
            Text("Godstone LabMesh").font(.headline)
            Text("EXPERIMENTAL / NONSHIPPING").font(.caption)
            Text("profile: \(LabMeshApp.profile)")
            Text("bundle: \(LabMeshApp.applicationId)")
            Text("link layer ready: \(LabMeshApp.readiness().iosLinkLayerReady ? "true" : "false")")
        }
        .padding()
    }
}

File: android/labmesh/src/main/java/io/godstone/labmesh/LabMeshApp.kt
package io.godstone.labmesh

import io.godstone.mesh.lab.LabProfile
import io.godstone.mesh.lab.LabRuntime

// ---------------------------------------------------------------------------
// T54 -- the LabMesh Android target's application seam.
//
// This module carrieth the lab's OWN application identity (io.godstone.labmesh,
// declared in build.gradle.kts) and nothing else: the runtime it driveth is the
// canonical one, reached through `io.godstone.mesh.lab.LabRuntime`, which liveth
// in the nonshipping :mesh module. There is no lab-only copy of any authority,
// and no way to manufacture readiness from here.
// ---------------------------------------------------------------------------

/** The lab application's own marker, carrying the identity and profile a reader
 *  (and the profile gate) can inspect without starting a runtime. */
object LabMeshApp {
    /** This lab's application identity -- never the shipping one. */
    const val APPLICATION_ID: String = LabProfile.LAB_APPLICATION_ID

    /** The shipping identity the lab may never masquerade as. */
    const val SHIPPING_APPLICATION_ID: String = LabProfile.SHIPPING_APPLICATION_ID

    /** The profile this build carrieth. */
    const val PROFILE: String = LabProfile.NAME

    /** True iff this build is experimental. It is. */
    const val EXPERIMENTAL: Boolean = LabProfile.EXPERIMENTAL

    /** True iff this build can manufacture crypto readiness. It cannot. */
    const val MANUFACTURES_READINESS: Boolean = LabProfile.MANUFACTURES_READINESS

    /** Compose the real runtime the lab driveth (the default A <-> R <-> B chain). */
    fun compose(): LabRuntime = LabRuntime.compose()

    /** Compose the real runtime with the caller's own peer labels. */
    fun compose(labels: List<String>): LabRuntime = LabRuntime.compose(labels)

    /** The honest readiness statement (all platform fields false). */
    fun readiness() = LabRuntime.readinessStatement()
}
